text/0000-simd-infrastructure.md

```diff
@@ -0,0 +1,404 @@
+but doesn't fill everything out. The goal here is to provide the raw
+types and access to the raw instructions on each platform.
+
+(An earlier variant of this RFC was discussed as a
+[pre-RFC](https://internals.rust-lang.org/t/pre-rfc-simd-groundwork/2343).)
+
+## Where does this code go? Aka. why not in `std`?
+
+This RFC is focused on building stable, powerful SIMD functionality in
+external crates, not `std`.
+
+This makes it much easier to support functionality only "occasionally"
+available with Rust's preexisting `cfg` system. There's no way for
+`std` to conditionally provide an API based on the target features
+used for the final artifact. Building `std` in every configuration is
+certainly untenable. Hence, if it were to be in `std`, there would
+need to be some highly delayed `cfg` system to support that sort of
+conditional API exposure.
+
+With an external crate, we can leverage `cargo`'s existing build
+infrastructure: compiling with some target features will rebuild with
+those features enabled.
+
+
+# Detailed design
+
+The design comes in three parts, all on the path to stabilisation:
+
+- types (`feature(simd_basics)`)
+- operations (`feature(simd_basics)`)
+- platform detection (`feature(cfg_target_feature)`)
+
+The general idea is to avoid bad performance cliffs, so that an
+intrinsic call in Rust maps to preferably one CPU instruction, or, if
+not, the "optimal" sequence required to do the given operation
+anyway. This means exposing a *lot* of platform specific details,
+since platforms behave very differently: both across architecture
+families (x86, x86-64, ARM, MIPS, ...), and even within a family
+(x86-64's Skylake, Haswell, Nehalem, ...).
+
+There is definitely a common core of SIMD functionality shared across
+many platforms, but this RFC doesn't try to extract that, it is just
+building tools that can be wrapped into a more uniform API later.
+
+
+## Types
+
+There is a new attributes: `repr(simd)`.
+
+```rust
+#[repr(simd)]
+struct f32x4(f32, f32, f32, f32);
+
+#[repr(simd)]
+struct Simd2<T>(T, T);
+```
+
+The `simd` `repr` can be attached to a struct and will cause such a
+struct to be compiled to a SIMD vector. It can be generic, but it is
+required that any fully monomorphised instance of the type consist of
+only a single "primitive" type, repeated some number of times. Types
+are flattened, so, for `struct Bar(u64);`, `Simd2<Bar>` has the same
+representation as `Simd2<u64>`.
+
+The `repr(simd)` may not enforce that any trait bounds exists/does the
+right thing at the type checking level for generic `repr(simd)`
+types. As such, it will be possible to get the code-generator to error
+out (ala the old `transmute` size errors), however, this shouldn't
+cause problems in practice: libraries wrapping this functionality
+would layer type-safety on top (i.e. generic `repr(simd)` types would
+use some `unsafe` trait as a bound that is designed to only be
+implemented by types that will work).
```

Comment:
> If Simd4<T> isn't exposed outside of the library, it's just an implementation convenience; if it is exposed, it's encouraging people to write code that gives weird errors

I also really really dislike monomorphisation errors, however, I don't think this is case too problematic. In practice, the generics will be bounded by traits, and users opt-in to getting a weird error by implementing the trait. I imagine that almost all instances of monomorphisation errors will be caught immediately: an implementation of the trait will presumably accompany uses of the type in SIMD vectors, resulting in the errors.

>  Using unsafe doesn't really improve anything (that's memory safety, not crash-the-compiler safety), and Rust doesn't have sealed traits at the moment.

One can seal traits via privacy. It's not sealing in a way that the compiler will reason about for type-system purposes, but it is enough for this. However, I'm not sure it makes sense.

> I think Rust's trait system is powerful enough that you could come up with a reasonable SIMD library without going down that path

This is very unclear to me. I suspect there will be a large amount of duplicated code, exponentially much, in fact: every operation that works on, say, 4-element vectors needs to be duplicated across every concrete type with that shape, e.g. `Simd4<i32>`, `Simd4<u32>`, `Simd4<f32>`, `Simd4<bool32>` definitely exist (and possibly more).

I'm very happy to be proved wrong, and this RFC is expressly designed to allow fairly free experimentation for higher level APIs. Also, it will remain unstable for a while, while we tweak (or, massively-change, if necessary) it.

One approach that I've discussed a bit is to "eventually" require some sort of trait bound on any generic `repr(simd)`'s, however, we don't know what this trait/traits should look like yet, and getting freeform experience will help guide that decision.

> actually, a user could even implement SIMD traits for types that aren't repr(simd).

This seems like something that's perfectly possible even with generic `repr(simd)`. (I'm missing the point.)